### PR TITLE
feat(review): reviewed oversized pull requests in batches instead of skipping them

### PR DESCRIPTION
**Pull request by rios0rios0**
## Problem

A pull request whose assembled prompt exceeds the AI model's context window got **no review at all** — only a notice telling the author to split it up:

> ⚠️ **Code Guru review couldn't run — this pull request is too large for the AI model's context window.**
> It changes 100 files (~3.7 MB of diff), which is more than the AI reviewer can read in a single pass, so no review was posted.

That is exactly backwards: the largest changes are the ones that most need a reviewer, and they are the ones the bot silently walked away from.

## What changed

When the backend reports a context-window overflow, the review is now **split into batches** instead of abandoned:

1. The pull request gets a `⏳ **Code Guru is reviewing this PR in batches.**` notice carrying the same scale figures, reframed around what actually happens — the review **is** coming, it just takes several times longer, so nobody reads the delay as a crashed bot and merges early.
2. The files are split into chunks that fit the model, reviewed one after another, and merged into a single review that flows through the normal post pipeline (stale/duplicate filters, completion annotation, native vote).

### Design notes

- **Batch size is discovered, not configured.** The first budget is derived from the token figures in the backend's own error (`prompt is too long: N tokens > M maximum` on Anthropic, `maximum context length is N tokens … resulted in M tokens` on OpenAI) scaled `0.8x` for the fixed non-diff prompt overhead, and halves on any batch that still overflows. It is clamped into `[total/maxBatches, total/2]` so a misread error cannot collapse the budget and destroy coverage. This adapts to the model, the 1M beta flag, and how much of the window the rules and guidelines already consume — none of which the bot can know up front.
- **Each batch's prompt states that it holds only a slice** (`ReviewRequest.Batch` → a `PARTIAL REVIEW` section). Without it, a sliced diff reliably produces three bogus findings about the parts the model cannot see: "this is never called", "the new flag has no tests", "the change is incomplete". The section collapses to nothing on a whole-PR request, so the **prompt no-drift invariant holds** — normal reviews are byte-for-byte unchanged.
- **Merging:** union of the comments, most severe verdict wins (`request_changes` > `comment` > `approve`), bounded join of the per-batch summaries. A run that could not read every file **never returns `approve`** — an approval claims the whole change was examined.
- **Gaps are reported, never hidden.** A file too large to review even alone, a failed batch, or a run that hits the batch cap is named in the merged summary. If *not one* batch succeeds, the historical "too large" annotation still fires. A single-file overflow skips batching entirely rather than promising a run it cannot deliver.
- **Re-review path:** each batch is shown only the prior threads anchored to its own files, and its batch-local `T1`/`T2` resolution ids are re-keyed to the run-global numbering — otherwise batch 2's `T1` would reply "fixed" on batch 1's thread and auto-close it.
- **Cancellation** is checked before every batch: a batched run is the longest-lived work the bot does, so it is the most likely to still be burning model calls on a pull request that was merged ten minutes ago.

### Configuration

| YAML key | Env var | Default |
|---|---|---|
| `ai.batch_large_reviews` | `CODE_GURU_AI_BATCH_LARGE_REVIEWS` | tri-state, **on** |
| `ai.max_review_batches` | `CODE_GURU_AI_MAX_REVIEW_BATCHES` | `20` |

Batching costs one model call per batch, so `batch_large_reviews: false` restores the previous give-up behaviour for spend-capped deployments.

## Tests

New `review_batches_test.go` drives the whole fallback against a hand-rolled double that models a finite context window (behavioural, not call-scripted — the planner's job is to *discover* a size that fits, so a call-count script would pass even if it never shrank anything). Covers: full coverage across batches, the shrink ladder converging, an unreviewable file being recorded while the rest still review, the never-approve-with-gaps rule, verdict merging, the batch cap, total failure surfacing the sentinel, cancellation, conversation partitioning + id re-keying, budget sizing, and both Execute-level paths (batched success, and the opt-out falling back to the too-large annotation). Plus prompt-section, notice-body, overage-parser, and settings-resolver tests.

`make lint`, `make test`, and `make sast` all pass.

## Docs

`CHANGELOG.md`, `README.md` (rewrote the "Large pull requests" section + config table/sample), `CLAUDE.md` (feature catalog, new batched-review flow, config table, invariants, ahead-of-peers list), and `.github/copilot-instructions.md`.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: internal/domain/commands/review_batches_test.go [hunk 1]
```diff
@@ -0,0 +1,869 @@
...
+// Hand-rolled rather than driven by a mock library (banned by the testing
+// standard) and deliberately behavioural rather than call-scripted: the
+// batch planner's whole job is to discover a size that fits, so a double
+// that answers by SIZE is what actually exercises it. A call-count script
+// would pass even if the planner never shrank anything.
+type windowedAIReviewer struct {
+	windowBytes int
+	// requests records every request the planner issued, in order, so a
+	// test can assert coverage (every file reviewed exactly once), the
+	// batch framing, and the conversation partitioning.
```
[the remark this answers is not in the document]
Comment (rios0rios0) [reply internal/domain/commands/review_batches_test.go:39]: Addressed in `661dda0`. Added a `successfulCalls` counter, incremented only after the window check passes, and indexed `verdicts`/`summaries`/`resolutions` off it — the double now matches its documented "Nth successful call" contract.

File: internal/domain/commands/review_batches_test.go [hunk 2]
```diff
@@ -0,0 +1,869 @@
...
+		weight += len(diff.Diff)
+	}
+	if weight > r.windowBytes {
+		return nil, fmt.Errorf(
+			"%w (windowed: prompt is too long: %d tokens > %d maximum)",
+			support.ErrContextWindowExceeded, weight, r.windowBytes,
+		)
+	}
+
+	call := len(r.requests)
```
[the remark this answers is not in the document]
Comment (rios0rios0) [reply internal/domain/commands/review_batches_test.go]: Addressed in `661dda0`. `call` now reads from the new `successfulCalls` counter rather than `len(r.requests)`, so scripted verdicts stay pinned to the batches that were actually reviewed. Also added a regression test that forces the planner to spend an overflow probe before its first successful batch — it fails under the old indexing (the run came back `approve` instead of the scripted `request_changes`), so the contract is now enforced rather than just documented.
